src/docmanager/context.ts

```diff
@@ -0,0 +1,511 @@
+  }
+
+  /**
+   * Create a new context.
+   */
+  createNew(path: string, model: IDocumentModel, options: IModelFactoryOptions,
+    contents: IContentsModel): string {
+    let context = new Context(this);
+    this._contexts[context.id] = {
+      context,
+      path,
+      model,
+      modelName: options.name,
+      opts: options.contentsOptions,
+      contentsModel: this._copyContentsModel(contents),
+      session: null
+    };
+    return context.id;
+  }
+
+  /**
+   * Get a context for a given path and model name.
+   */
+  findContext(path: string, modelName: string): string {
+    for (let id in this._contexts) {
+      let contextEx = this._contexts[id];
+      if (contextEx.path === path && contextEx.modelName === modelName) {
+        return id;
+      }
+    }
+  }
+
+  /**
+   * Find a context by path.
+   */
+  getIdsForPath(path: string): string[] {
+    let ids: string[] = [];
+    for (let id in this._contexts) {
+      if (this._contexts[id].path === path) {
+        ids.push(id);
+      }
+    }
+    return ids;
+  }
+
+  /**
+   * Get a context by id.
+   */
+  getContext(id: string): IDocumentContext {
+    return this._contexts[id].context;
+  }
+
+  /**
+   * Get the model associated with a context.
+   */
+  getModel(id: string): IDocumentModel {
+    return this._contexts[id].model;
+  }
+
+  /**
+   * Remove a context.
+   */
+  removeContext(id: string): INotebookSession {
+    let contextEx = this._contexts[id];
+    contextEx.model.dispose();
+    contextEx.context.dispose();
+    delete this._contexts[id];
+    return contextEx.session;
+  }
+
+  /**
+   * Get the current kernel associated with a document.
+   */
+  getKernel(id: string): IKernel {
+    let session = this._contexts[id].session;
+    return session ? session.kernel : null;
+  }
+
+  /**
+   * Get the current path associated with a document.
+   */
+  getPath(id: string): string {
+    return this._contexts[id].path;
+  }
+
+  /**
+   * Get the current contents model associated with a document.
+   */
+  getContentsModel(id: string): IContentsModel {
+    return this._contexts[id].contentsModel;
+  }
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(id: string, options: IKernelId): Promise<IKernel> {
+    let contextEx = this._contexts[id];
+    let session = contextEx.session;
+    if (!session) {
+      let path = contextEx.path;
+      let sOptions = {
+        notebook: { path },
+        kernel: { options }
+      }
+      return this._startSession(id, sOptions);
+    } else {
+      return session.changeKernel(options);
+    }
+  }
+
+  /**
+   * Update the path of an open document.
+   *
+   * @param id - The id of the context.
+   *
+   * @param newPath - The new path.
+   */
+  rename(id: string, newPath: string): Promise<void> {
+    let contextEx = this._contexts[id];
+    let session = contextEx.session;
+    if (session) {
+      return session.renameNotebook(newPath);
```

Comment:
Renaming the notebook triggers a `notebookPathChanged` on all of the client sessions, which will be handled by `_startSession`.